File: vi/codes/ruby/chapter_divide_and_conquer/build_tree.rb
=begin
File: build_tree.rb
Created Time: 2024-05-13
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/tree_node'
require_relative '../utils/print_util'

### Xây dựng cây nhị phân: chia để trị ###
def dfs(preorder, inorder_map, i, l, r)
  # Kết thúc khi khoảng của cây con rỗng
  return if r - l < 0

  # Khởi tạo nút gốc
  root = TreeNode.new(preorder[i])
  # Truy vấn m để chia cây con trái và phải
  m = inorder_map[preorder[i]]
  # Bài toán con: xây dựng cây con trái
  root.left = dfs(preorder, inorder_map, i + 1, l, m - 1)
  # Bài toán con: xây dựng cây con phải
  root.right = dfs(preorder, inorder_map, i + 1 + m - l, m + 1, r)

  # Trả về nút gốc
  root
end

### Xây dựng cây nhị phân ###
def build_tree(preorder, inorder)
  # Khởi tạo bảng băm, lưu ánh xạ từ phần tử duyệt giữa đến chỉ số
  inorder_map = {}
  inorder.each_with_index { |val, i| inorder_map[val] = i }
  dfs(preorder, inorder_map, 0, 0, inorder.length - 1)
end

### Mã chương trình chính ###
if __FILE__ == $0
  preorder = [3, 9, 2, 1, 7]
  inorder = [9, 3, 1, 2, 7]
  puts "Pre-order traversal = #{preorder}"
  puts "In-order traversal = #{inorder}"

  root = build_tree(preorder, inorder)
  puts "The constructed binary tree is:"
  print_tree(root)
end

File: vi/codes/dart/chapter_divide_and_conquer/build_tree.dart
/**
 * File: build_tree.dart
 * Created Time: 2023-08-10
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

import '../utils/print_util.dart';
import '../utils/tree_node.dart';

/* Xây dựng cây nhị phân: chia để trị */
TreeNode? dfs(
  List<int> preorder,
  Map<int, int> inorderMap,
  int i,
  int l,
  int r,
) {
  // Dừng lại khi khoảng của cây con rỗng
  if (r - l < 0) {
    return null;
  }
  // Khởi tạo nút gốc
  TreeNode? root = TreeNode(preorder[i]);
  // Truy vấn m để phân chia cây con trái và phải
  int m = inorderMap[preorder[i]]!;
  // Bài toán con: xây dựng cây con trái
  root.left = dfs(preorder, inorderMap, i + 1, l, m - 1);
  // Bài toán con: xây dựng cây con phải
  root.right = dfs(preorder, inorderMap, i + 1 + m - l, m + 1, r);
  // Trả về nút gốc
  return root;
}

/* Xây dựng cây nhị phân */
TreeNode? buildTree(List<int> preorder, List<int> inorder) {
  // Khởi tạo bảng băm, lưu ánh xạ từ phần tử duyệt giữa đến chỉ số
  Map<int, int> inorderMap = {};
  for (int i = 0; i < inorder.length; i++) {
    inorderMap[inorder[i]] = i;
  }
  TreeNode? root = dfs(preorder, inorderMap, 0, 0, inorder.length - 1);
  return root;
}

/* Driver Code */
void main() {
  List<int> preorder = [3, 9, 2, 1, 7];
  List<int> inorder = [9, 3, 1, 2, 7];
  print("Pre-order traversal = $preorder");
  print("In-order traversal = $inorder");

  TreeNode? root = buildTree(preorder, inorder);
  print("The constructed binary tree is:");
  printTree(root!);
}
